system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -5782,6 +5849,283 @@ class CLdapClient : public CInterface, implements ILdapClient
+    {
+        StringArray currFiles;
+        StringArray currCols;
+        queryViewColumns(viewName, currFiles, currCols);
+        unsigned vCount = currFiles.ordinality();
+        assertex(vCount == currCols.ordinality());
+
+        unsigned len = files.ordinality();
+        assertex(len == columns.ordinality());
+        bool changed = false;
+        for(unsigned idx = 0; idx < len; idx++)
+        {
+            bool isDup = false;
+            for (unsigned vIdx = 0; vIdx < vCount; vIdx++)//look for dups
+            {
+                if (0 == stricmp(files.item(idx), currFiles.item(vIdx)) &&
+                    0 == stricmp(columns.item(idx), currCols.item(vIdx)))
+                {
+                    isDup = true;//skip duplicate entry
+                    break;
+                }
+            }
+
+            if (!isDup)
+            {
+                currFiles.append(files.item(idx));
+                currCols.append(columns.item(idx));
+                changed = true;
+            }
+        }
+
+        if (!changed)
+        {
+            throw MakeStringException(-1, "Specified columns already exist in view");
+        }
+
+        ///build description buffer containing one or more !!!lfn!!col
+        StringBuffer description;
+        len = currFiles.ordinality();
+        for(unsigned idx = 0; idx < len; idx++)
+        {
+            description.appendf("!!%s!%s",currFiles.item(idx), currCols.item(idx));//use illegal LFN character as separators
+        }
+
+        updateViewContents(viewName, description.str());
+    }
+
+    void removeViewColumns(const char * viewName, StringArray & files, StringArray & columns)
+    {
+        StringArray currFiles;
+        StringArray currCols;
+        queryViewColumns(viewName, currFiles, currCols);
+
+        unsigned len = files.ordinality();
+        assertex(len == columns.ordinality());
+        bool changed = false;
+        for(unsigned idx = 0; idx < len; idx++)//for all pairs to be removed
+        {
+            unsigned len2 = currFiles.ordinality();
+            for(unsigned idx2 = 0; idx2 < len2; idx2++)
+            {
+                if (0 == stricmp(files.item(idx), currFiles.item(idx2)) &&
+                    0 == stricmp(columns.item(idx), currCols.item(idx2)))
+                {
+                    currFiles.remove(idx2);
+                    currCols.remove(idx2);
+                    changed = true;
+                    break;
+                }
+            }
+        }
+
+        if (!changed)
+        {
+            throw MakeStringException(-1, "Specified columns do not exist in view");
+        }
+
+        ///build description buffer containing one or more !!!lfn!!col
+        StringBuffer description;
+        len = currFiles.ordinality();
+        for(unsigned idx = 0; idx < len; idx++)
+        {
+            description.appendf("!!%s!%s",currFiles.item(idx), currCols.item(idx));//use illegal LFN character as separators
```

Comment:
We cannot use '!' as a delimiter because '!' is a LEGAL LFN character.

We probably should use '|' instead, which is a definite illegal LFN character.